gdk4/src/event.rs

```diff
@@ -0,0 +1,527 @@
+        unsafe { from_glib(ffi::gdk_event_get_event_type(self.to_glib_none().0)) }
+    }
+
+    pub fn get_history(&self) -> Vec<TimeCoord> {
+        unsafe {
+            let mut out_n_coords = mem::MaybeUninit::uninit();
+            let ret = FromGlibContainer::from_glib_container_num(
+                ffi::gdk_event_get_history(self.to_glib_none().0, out_n_coords.as_mut_ptr()),
+                out_n_coords.assume_init() as usize,
+            );
+            ret
+        }
+    }
+
+    pub fn get_modifier_state(&self) -> ModifierType {
+        unsafe { from_glib(ffi::gdk_event_get_modifier_state(self.to_glib_none().0)) }
+    }
+
+    pub fn get_pointer_emulated(&self) -> bool {
+        unsafe { from_glib(ffi::gdk_event_get_pointer_emulated(self.to_glib_none().0)) }
+    }
+
+    pub fn get_position(&self) -> Option<(f64, f64)> {
+        unsafe {
+            let mut x = mem::MaybeUninit::uninit();
+            let mut y = mem::MaybeUninit::uninit();
+            let ret = from_glib(ffi::gdk_event_get_position(
+                self.to_glib_none().0,
+                x.as_mut_ptr(),
+                y.as_mut_ptr(),
+            ));
+            let x = x.assume_init();
+            let y = y.assume_init();
+            if ret {
+                Some((x, y))
+            } else {
+                None
+            }
+        }
+    }
+
+    pub fn get_seat(&self) -> Option<Seat> {
+        unsafe { from_glib_none(ffi::gdk_event_get_seat(self.to_glib_none().0)) }
+    }
+
+    pub fn get_surface(&self) -> Option<Surface> {
+        unsafe { from_glib_none(ffi::gdk_event_get_surface(self.to_glib_none().0)) }
+    }
+
+    pub fn get_time(&self) -> u32 {
+        unsafe { ffi::gdk_event_get_time(self.to_glib_none().0) }
+    }
+
+    pub fn triggers_context_menu(&self) -> bool {
+        unsafe { from_glib(ffi::gdk_event_triggers_context_menu(self.to_glib_none().0)) }
+    }
+}
+
+impl fmt::Display for Event {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.write_str("Event")
+    }
+}
+
+pub trait EventTypeTrait: StaticType + FromGlibPtrFull<*mut ffi::GdkEvent> + 'static {}
+
+macro_rules! define_event {
+    ($rust_type:ident, $ffi_type:path, $rust_type_offset:expr) => {
+        // Can't use get_type here as this is not a boxed type but another fundamental type
+        glib::wrapper! {
+            #[derive(Debug)]
+            pub struct $rust_type(Shared<$ffi_type>);
+
+            match fn {
+
+                ref => |ptr| ffi::gdk_event_ref(ptr as *mut ffi::GdkEvent) as *mut $ffi_type,
+                unref => |ptr| ffi::gdk_event_unref(ptr as *mut ffi::GdkEvent),
+            }
+        }
+
+        impl StaticType for $rust_type {
+            fn static_type() -> Type {
+                unsafe { from_glib(*glib::g_param_spec_types.add($rust_type_offset)) }
+            }
+        }
+
+        unsafe impl Send for $rust_type {}
+        unsafe impl Sync for $rust_type {}
```

Comment:
I suppose it's fine to implement those for an event? :) I wonder how can I find if a type can be Sync/Send?

Reply:
Probably not, nothing in GTK is thread-safe.